### PR TITLE
Enhancement/29 port resources

### LINKED ISSUE #27: Provide port injection for arguments
E.g.
`mesos.command.command=./webserver.sh`
`mesos.command.arguments=--port=$MESOS_PORT_0`
`mesos.docker.command=kibana`
`mesos.docker.arguments=--port=$MESOS_PORT_0`

Current thinking is to get the starter to create standardized env vars for the created ports and inject them into the container/command. Then the application can use the env vars.

Martin also suggested SPEL.

### PR DESCRIPTION
**Pull request by philwinder**
Working port injection.  …
- Refactor docker configuration options into DockerConfigProperties
- Add new PortRequirement
- Add port parser to parse different types of ports (fixed, unprivileged, privileged)
- No longer publish all ports (will be published in Host mode)
- Add properties to expose ports through env vars
- Host/Bridge mode.

Working host example:
```
spring.application.name=kibana
mesos.framework.name=kibana
mesos.resources.scale=3
mesos.resources.cpus=0.5
mesos.resources.mem=256
mesos.command=kibana --port=$UI_5061 --elasticsearch http://172.17.0.2:9200
mesos.resources.ports=true
mesos.resources.ports.UI_5061=ANY
mesos.docker.network=HOST
mesos.docker.image=kibana:latest
```
Working bridge example:
```
spring.application.name=kibana
mesos.framework.name=kibana
mesos.resources.scale=3
mesos.resources.cpus=0.5
mesos.resources.mem=256
mesos.command=kibana --port=$UI_5061 --elasticsearch http://172.17.0.2:9200
mesos.resources.ports=true
mesos.resources.ports.UI_5061=ANY
mesos.docker.parameter.expose=$UI_5061
mesos.docker.network=BRIDGE
mesos.docker.image=kibana:latest
```
Working jar mode:
```
spring.application.name=kibana
mesos.framework.name=kibana
mesos.resources.scale=3
mesos.resources.cpus=0.5
mesos.resources.mem=256
mesos.command=curl -sS -q -o kibana.tar.gz https://download.elastic.co/kibana/kibana/kibana-4.4.1-linux-x64.tar.gz > /dev/null ; tar -zxf kibana.tar.gz ; cd kibana-* ; bin/kibana --port=$UI_5061 --elasticsearch http://172.17.0.2:9200
mesos.resources.ports=true
mesos.resources.ports.UI_5061=ANY
```
Fixes #27 and #29

### CODE REVIEW

**Comment by mwl** [comment]
@philwinder Looks like there's a merge conflict with Master

File: spring-boot-starter-mesos/src/test/java/com/containersolutions/mesos/scheduler/requirements/ports/PortUtil.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
+package com.containersolutions.mesos.scheduler.requirements.ports;
+
+import org.apache.mesos.Protos;
+
+import static com.containersolutions.mesos.config.autoconfigure.MesosSchedulerConfiguration.MESOS_PORTS;
+
+public class PortUtil {
```
Comment (mwl) [inline spring-boot-starter-mesos/src/test/java/com/containersolutions/mesos/scheduler/requirements/ports/PortUtil.java]: There's already a `com.containersolutions.mesos.TestHelper` class, where these small helper method will fit in very nicely.

**Comment by philwinder** [comment]
There is a problem with this line:
https://github.com/ContainerSolutions/mesos-starter/blob/master/spring-boot-starter-mesos/src/main/java/com/containersolutions/mesos/config/autoconfigure/MesosSchedulerConfiguration.java#L169

A map isn't considered to be a present property. So when a user applies a port requirement, the port requirement isn't instantiated.

**Comment by mwl** [comment]
Didn't you rename it to plural `ports`?

**Comment by mwl** [comment]
> Working bridge example:
> 
> spring.application.name=kibana
> mesos.framework.name=kibana
> mesos.resources.scale=3
> mesos.resources.cpus=0.5
> mesos.resources.mem=256
> mesos.command=kibana --port=$UI_5061 --elasticsearch http://172.17.0.2:9200
> mesos.resources.ports=true
> mesos.resources.ports.UI_5061=ANY
> mesos.docker.parameter.expose=$UI_5061
> mesos.docker.network=BRIDGE
> mesos.docker.image=kibana:latest

Documentation seems wrong? Value of `$UI_5061` is the port from the Mesos Resources (i.e. 31000). To me it makes no sense to start Kibana on that port, and expose it when in Bridge mode. Normally you never touch the container ports when in Bridge mode. You just map them to `$UI_5061` on the host.

File: spring-boot-starter-mesos/src/main/java/com/containersolutions/mesos/scheduler/TaskInfoFactoryDocker.java [hunk 2]
```diff
@@ -33,39 +28,55 @@
...
-        Iterator<String> portsIterator = mesosConfig.getResources().getPort().iterator();
-        return resources.stream()
-                .filter(Protos.Resource::hasRanges)
-                .filter(resource -> resource.getName().equals("ports"))
-                .flatMap(resource -> resource.getRanges().getRangeList().stream())
-                .flatMapToLong(range -> LongStream.rangeClosed(range.getBegin(), range.getEnd()))
-                .limit(mesosConfig.getResources().getPort().size())
-                .mapToObj(hostPort -> Protos.ContainerInfo.DockerInfo.PortMapping.newBuilder().setHostPort((int) hostPort).setContainerPort(Integer.parseInt(portsIterator.next())).build())
-                .peek(portMapping -> logger.debug("Mapped host=" + portMapping.getHostPort() + "=>" + portMapping.getContainerPort()))
+        return dockerParameters.entrySet().stream()
```
Comment (mwl) [inline spring-boot-starter-mesos/src/main/java/com/containersolutions/mesos/scheduler/TaskInfoFactoryDocker.java]: I can't see what we need this for. Only example given is to expose ports in Bridge mode, but that should already have been done in the Docker Image

**Comment by mwl** [comment]
Solved in #39

**Pull request closed without merging**